src/app/views/query-runner/query-input/AutoComplete.tsx

```diff
@@ -0,0 +1,344 @@
+    if (this.autoCompleteRef && this.autoCompleteRef.current) {
+      this.autoCompleteRef.current.focus();
+    }
+  }
+
+  public onChange = (e: any) => {
+    const { suggestions, showSuggestions, userInput: previousUserInput, compare } = this.state;
+    const userInput = e.target.value;
+
+    this.props.contentChanged(userInput);
+
+    this.setState({
+      userInput
+    });
+
+    if (showSuggestions && suggestions.length) {
+      this.filterSuggestions(userInput, previousUserInput, compare, suggestions);
+    }
+    this.initialiseAutoComplete(userInput);
+  };
+
+  public onClick = (e: any) => {
+    const selected = e.currentTarget.innerText;
+    this.appendSuggestionToUrl(selected);
+  };
+
+  private initialiseAutoComplete = (url: string) => {
+    const lastCharacter = url.substring(url.length - 1);
+    switch (lastCharacter) {
+      case '/':
+        this.requestForAutocompleteOptions(url);
+        break;
+
+      case '?':
+        this.requestForAutocompleteOptions(url);
+        break;
+
+      case '=':
+
+        if (url.includes('?$')) {
+          this.getParameterEnums(url);
+        }
+
+        break;
+
+      case ',':
+        this.getParameterEnums(url);
+        break;
+
+      case '&':
+        this.getQueryParameters();
+        break;
+
+      default:
+        break;
+    }
+  }
+
+  public onKeyDown = (e: any) => {
+    const { activeSuggestion, filteredSuggestions, showSuggestions } = this.state;
+
+    switch (e.keyCode) {
+      case KeyCodes.enter:
+        if (showSuggestions) {
+          const selected = filteredSuggestions[activeSuggestion];
+          this.appendSuggestionToUrl(selected);
+        } else {
+          this.props.runQuery();
+        }
+        break;
+
+      case KeyCodes.up:
+        if (activeSuggestion === 0) {
+          return;
+        }
+        this.setState({ activeSuggestion: activeSuggestion - 1 });
+        break;
+
+      case KeyCodes.down:
+        if (activeSuggestion - 1 === filteredSuggestions.length) {
+          return;
+        }
+        this.setState({ activeSuggestion: activeSuggestion + 1 });
+        break;
+
+      default:
+        break;
+    }
+
+  };
+
+  public displayLinkOptions = () => {
+    const { autoCompleteOptions, sampleQuery: { selectedVerb } } = this.props;
+    if (autoCompleteOptions) {
+      const suggestions: string[] = [];
+      const parameters = autoCompleteOptions.parameters;
+      if (parameters) {
+        const parametersWithVerb = parameters.find(parameter => parameter.verb === selectedVerb.toLowerCase());
+        if (parametersWithVerb) {
+          parametersWithVerb.links.forEach((value: string) => {
+            suggestions.push(value);
+          });
+
+          this.setState({
+            filteredSuggestions: suggestions,
+            suggestions,
+            showSuggestions: true,
+            compare: ''
+          });
+        }
+      }
+    }
+  }
+
+  public getQueryParameters = () => {
```

Comment:
Function body is exactly the same as displayLinkOptions except for the part code pushes into suggestions. I think duplicated logic in these two functions can be abstracted out as functions:

```
getParametersWithVerb()
{
// code that returns parametersWithVerb
}

setSuggestions(suggestions: string[])
{
  this.setState({
    filteredSuggestions: suggestions,
    suggestions,
    showSuggestions: true,
    compare: ''
  });
}

public getQueryParameters = () => {
  const parametersWithVerb = getParametersWithVerb();
  if (!parametersWithVerb)
  {
    return;
  }
  setSuggestions(parametersWithVerb.values.map(value => value.name));
}

public displayLinkOptions = () => {
  const parametersWithVerb = getParametersWithVerb();
  if (!parametersWithVerb)
  {
    return;
  }
  setSuggestions(parametersWithVerb.links);
}
```